Support reading multiple bytes in escapeProxy

bamarni opened the pull request:
Currently, the escapeProxy works under the assumption that the
underlying reader will always return 1 byte at a time. Even though this
is usually true, it is not always the case, for example when using a pty
and writing multiple bytes to the master before flushing it.

In such cases the proxy reader doesn't work properly. For example with
an escape sequence being `ctrl-p,ctrl-q`, when the underlying reader
returns `ctrl-p,ctrl-q` at once, the escape sequence isn't detected.

A second issue occurs when there is a partial escape currently being
read, if on the next read the len of the passed buffer in argument is
not big enough, it will cause a panic. For example, when the escape sequence
is `ctrl-p,ctrl-q`, and the caller is passing a []byte of len 1 as argument to read.
And they press `ctrl-p`, then something else than `ctrl-q`. 

This updates the reader to support this use-cases and adds unit tests.

I'd still like to test more cases to make sure it's working fine, but I'm first
opening the PR to see if it is an acceptable patch.

thaJeztah commented on pkg/term/proxy.go:
```diff
@@ -31,48 +32,54 @@ func NewEscapeProxy(r io.Reader, escapeKeys []byte) io.Reader {
+			if r.escapeKeyPos == len(r.escapeKeys) {
+				n = i + 1
+				err = EscapeError{}
+				break
```
Wondering if it'd be clearer to return early here (possibly at the cost of some slight duplication)

bamarni replied:
Here I think it could be replaced with:
``` go
n = i + 1 - r.escapeKeyPos
if n < 0 {
    n = 0
}
return n, EscapeError{}
```

Let me know if that looks better and I'll try it out.

dims replied:
yes please! +1 to "return early"

dims replied:
may be in a follow up PR is fine.

thaJeztah commented on pkg/term/proxy.go:
```diff
@@ -31,48 +32,54 @@ func NewEscapeProxy(r io.Reader, escapeKeys []byte) io.Reader {
 
-func (r *escapeProxy) Read(buf []byte) (int, error) {
-	nr, err := r.r.Read(buf)
+func (r *escapeProxy) Read(buf []byte) (n int, err error) {
```
Not a huge fan of named output variables; it's easy to inadvertently mask them and introduce bugs; perhaps we can slightly refactor the code to not require them

bamarni replied:
I'd have to make variable declarations at the beginning of the function, this wouldn't look better IMO, I don't see an alternative with the current implementation.

dims replied:
Hmm, 🤞 that we won't have to touch this again! so ok for named output variable here

thaJeztah commented on pkg/term/proxy_test.go:
```diff
@@ -112,4 +112,53 @@ func TestEscapeProxyRead(t *testing.T) {
 	assert.Equal(t, nr, len(keys), fmt.Sprintf("nr should be equal to %d", len(keys)))
 	assert.DeepEqual(t, keys, buf)
+
+	escapeKeys, _ = ToBytes("ctrl-c,ctrl-z")
```
This test becomes really unwieldy with all the test-cases (some of which look to depend on values set by previous test-cases).

Looks like they all _slightly_ differ (so we cannot easily change it into test-table, but we can use subtests to group them into logical test-cases, which would make this test easier to read

thaJeztah replied:
Went ahead and did some refactoring; https://github.com/moby/moby/pull/39800 (you can update the new tests to use the same / similar pattern)

bamarni replied:
Sounds good, should I cherry-pick your commit in my PR, or do you expect it to be merged first and then I'll rebase?

thaJeztah replied:
Hope it will be merged soon, but you could already modify the new tests to match

bamarni replied:
Great, I've pulled your changes and will rebase once they're merged.

thaJeztah reviewed (changes requested):
Thanks! Did a cursory review, and left some comments/suggestions

thaJeztah commented:
Sorry for the delay; https://github.com/moby/moby/pull/39800 was merged, so you can rebase this PR 🤗

bamarni commented:
@thaJeztah : awesome :tada: it's rebased now!

bamarni commented:
The build failed:

```
[2019-09-07T21:47:09.271Z] --- FAIL: TestLogBlocking (0.04s)
[2019-09-07T21:47:09.271Z]     cloudwatchlogs_test.go:313: Expected to be able to read from stream.messages but was unable to
[2019-09-07T21:47:09.271Z] time="2019-09-07T21:47:06Z" level=error msg=Error
[2019-09-07T21:47:09.271Z] time="2019-09-07T21:47:06Z" level=error msg="Failed to put log events" errorCode=InvalidSequenceTokenException logGroupName=groupName logStreamName=streamName message="use token token" origError="<nil>"
[2019-09-07T21:47:09.271Z] time="2019-09-07T21:47:06Z" level=error msg="Failed to put log events" errorCode=DataAlreadyAcceptedException logGroupName=groupName logStreamName=streamName message="use token token" origError="<nil>"
[2019-09-07T21:47:09.272Z] time="2019-09-07T21:47:06Z" level=info msg="Data already accepted, ignoring error" errorCode=DataAlreadyAcceptedException logGroupName=groupName logStreamName=streamName message="use token token"
[2019-09-07T21:47:09.272Z] FAIL
```

See https://ci.docker.com/public/blue/organizations/jenkins/moby/detail/PR-39728/4/pipeline

I'll try to re-run.

bamarni commented:
@thaJeztah : I've rebased since #39887 was merged and CI passed :rocket:

bamarni commented:
@thaJeztah @runcom : I've rebased with the PR improving the tests and CI is now passing, is there anything else missing here?

bamarni commented:
@thaJeztah : is this PR still on your radar?

thaJeztah reviewed (approved):
ah, whoops, no it wasn't 😅 

LGTM

thaJeztah commented:
ping @vdemeester PTAL, I think you're familiar with this area

bamarni commented:
@vdemeester any news on this one? :angel:

bamarni commented:
Hey @thaJeztah @vdemeester, can you help with this one? It's been sitting here for a while.

dims commented on pkg/term/proxy_test.go:
```diff
@@ -148,4 +160,49 @@ func TestEscapeProxyRead(t *testing.T) {
 		assert.DeepEqual(t, keys, buf)
 	})
 
+	t.Run("a,b,c,d escape key, keys [a,b],[c,d]", func(t *testing.T) {
```
thanks for the additional tests. helps understand the problem better.

dims commented:
LGTM

thanks @bamarni

bamarni commented:
@dims : so is it good to go, or should I apply your comment about early return here?

dims commented:
@bamarni if you can quickly make the change for the early return, that would be good. I am fine if you want to do it in a follow up PR as this has been stuck forever

cpuguy83 commented:
If we are going to make this change, I'd like to simplify this down.
I feel like we can do this with a bufio.Reader and `ReadByte`.

I kinda started messing with this, it's not 100% yet, but PTAL:

```go
func (r *escapeProxy) Read(p []byte) (nr int, err error) {
	if len(r.escapeKeys) == 0 {
		// This causes tests to fail b/c they are expecting an error in this case.
		// I'm not sure that's correct... but in any case this is a place holder to prevent test panics
		return r.r.Read(p)
	}

	for i := 0; i < len(p); i++ {
		b, err := r.r.ReadByte()
		if err != nil {
			return nr, err
		}
		p[nr] = b

		if b != r.escapeKeys[r.escapeKeyPos] {
			r.escapeKeyPos = 0
			nr++
			continue
		}

		r.escapeKeyPos++
		if r.escapeKeyPos == len(r.escapeKeys) {
			return nr, EscapeError{}
		}
	}

	return nr, nil
}
```

bamarni committed:
Support reading multiple bytes in escapeProxy

Currently, the escapeProxy works under the assumption that the
underlying reader will always return 1 byte at a time. Even though this
is usually true, it is not always the case, for example when using a pty
and writing multiple bytes to the master before flushing it.

In such cases the proxy reader doesn't work properly. For example with
an escape sequence being `ctrl-p,ctrl-q`, when the underlying reader
returns `ctrl-p,ctrl-q` at once, the escape sequence isn't detected.

This updates the reader to support this use-case and adds unit tests.

Signed-off-by: Bilal Amarni <bilal.amarni@gmail.com>

bamarni commented:
@dims : I've updated the code as suggested.

@cpuguy83 : I'll try to check this out. One notable difference is that it with your sample it will make multiple read syscalls instead of 1, is that not a concern?

bamarni commented:
@cpuguy83 : Ah I see, so the bufio.Reader would read 4096 by default and keep it in memory, then we'd iterate one byte at a time.

bamarni commented:
@cpuguy83 : I've looked at your snippet: 

- regarding your comment about the test failure because it expects an error, this is due to a difference between `bytes.Reader` and `bufio.Reader`, one will return `io.EOF` when you pass it a p of length 0, while the error wouldn't. See https://play.golang.org/p/YU4NAr0W52o

- some other test failures are because the snippet doesn't handle the case where there were a previous call to the escape proxy reader which partially matched an escape sequence, at this point it returned no data. However, on a subsequent call, the proxy reader should not read and return data from the underlying reader straight away, rather, it should fill the p with the partial escape sequence previously matched.

It would require additional logic to fix these 2 (especially the 2nd point), and eventually I'm not sure that the resulting code would be necessarily simpler. Let me know what you think.

bamarni commented:
Hey @dims @cpuguy83, any news?

dims commented:
@bamarni waiting on @cpuguy83

AkihiroSuda commented:
@cpuguy83 PTAL

cpuguy83 reviewed (approved):
LGTM

Should we move this to github.com/moby/term?

dims commented:
LGTM

+1 to move this to moby/term

thaJeztah reviewed (approved):
LGTM, thanks!
